### PR TITLE
Add new savegame format

### LINKED ISSUE #277: Feature request: new savegame format
I'd like us to design a new savegame format that can easily maintain backwards compatibility in case we need to make changes to what gets saved (see #226, #261, and possibly more).

Arsunt's proposition was to use [TLV](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value) and I think it is sensible, however I think that it's OK for us not to maintain backwards compatibility with TombATI.

I would also like to suggest [BSON](https://bsonspec.org/spec.html).

I'd still like to support legacy format, but only loading.

### PR DESCRIPTION
**Pull request by rr-**
Resolves #277 

The new format uses zlib-compressed BSON. The game always saves in the new format, which uses a new filename convention. In case a given save slot has files in both formats available, the game prioritizes the new format upon loading. Loading old saves is still possible. In theory saving in the legacy format is also possible, but I advise against bringing it back.

Other things:
- made the format resilient. No more crashes when trying to load saves in the new format made for a mod or a custom level, instead there is a log message and the game exits to title.
- removed `GAMEINFO.savegame_buffer` which was frankly quite ugly. I still need info which save slot to load from upon game start, but it's a lot better as it decouples the game from the savegame structure by a great deal already.
- refactored JSON routines.
- refactored save and loading to cater for the new multi-format philosophy.
- added a Python script that can inspect the new savegame files; it is located in `scripts/inspect_save`.

One thing I'd like to investigate before we merge, provided everything is OK, is whether these snippets, copied from the legacy loader,
```
    size_t idx = (size_t)arm->frame_base - (size_t)g_AnimFrames;
```
are dependent on the size of the g_AnimFrames structure, or retrieve the index.

### CODE REVIEW

**Comment by rr-** [comment]
I also tested the BSON serializer and deserializer with valgrind. There should be no memory leaks or invalid writes. I didn't test the BSON savegame format with valgrind as this requires linking against the actual game internals, which is not possible on my system (yet).

**Comment by rr-** [comment]
The saves take around 2-3 KB on average :)

**Comment by rr-** [comment]
To make things easier on walkawayy @richard-l could you please help with testing? Please note that the new build requires overwriting the gameflow files.

[Tomb1Main-2.5-25-g159f396.zip](https://github.com/rr-/Tomb1Main/files/8008822/Tomb1Main-2.5-25-g159f396.zip)

**Review by walkawayy — COMMENTED** [review]
I did a skim and most of the changes look good but I didn't do much in game testing.

Are we aiming for new features in the new savegame format in this PR? Or is this just to get the new bson format to match legacy saves, and then we can add additional features (like we've seen in various issue) in the future?

File: bin/cfg/Tomb1Main_gameflow.json5 [hunk 1]
```diff
@@ -6,7 +6,8 @@
     "main_menu_picture": "data\\titleh.png",
 
     // path to the savegame file
-    "savegame_fmt": "saveati.%d",
+    "savegame_fmt_legacy": "saveati.%d",
+    "savegame_fmt_bson": "save_tr1_%02d.dat",
```
[the remark this answers is not in the document]
Comment (rr-) [reply bin/cfg/Tomb1Main_gameflow.json5:10]: That's issue #87 and I'd like to handle that in a separate PR.

**Comment by rr-** [comment]
No, the changes relevant to other tickets will be tackled in follow-up pull requests. This is only about adding an extensible format.

**Comment by Richard-L** [comment]
> To make things easier on walkawayy @Richard-L could you please help with testing? Please note that the new build requires overwriting the gameflow files since we introduced new settings there.
> 
> [Tomb1Main-2.5-25-g159f396.zip](https://github.com/rr-/Tomb1Main/files/8008822/Tomb1Main-2.5-25-g159f396.zip)

Looks very nice. So far can't find anything that seems wrong. I see the new saves replace the old saves in the list, but stay physically untouched on the hard drive. I think this is good practice. I also notice that new saves will correctly continue the save counter next to them, even if they are NSF. This looks very polished to me already.

Anything you think could be worth testing in particular?

**Comment by rr-** [comment]
Yeah I'm mostly looking for regressions around actual savegame behavior, such as:

- persisting ammo, guns, medpacks, key items, scions
- persisting kill count, timer, collected secrets
- not getting stationary enemies
- having dead enemies stay dead
- persisting Lara's equipped guns, animations
- while in a current
- while falling
- while underwater
- while on a water surface
- flipmaps on and flipmaps off
- behavior of Bacon Lara

More?

**Comment by rr-** [comment]
[Tomb1Main-2.5-28-g3371f97.zip](https://github.com/rr-/Tomb1Main/files/8009067/Tomb1Main-2.5-28-g3371f97.zip)

Fixed an issue with storing meshes, animation numbers and effects that made them dependent on the size of the memory structures we use to store these. This made the saves made with the previous build unable to load, so please delete them first.

Also optimized the savegame listing so that it does not parse the files 4 times, just 2 (once for listing and once for actual loading).

---

NB I'm not sure why GitHub does not show my recent changes,
```
* 3371f976 2022-02-05 savegame: optimize listing savegames               [rr- <dash@wind.garden>] (HEAD -> bson, origin/bson)
* 8086924e 2022-02-05 savegame: make indices not depend on struct sizes  [rr- <dash@wind.garden>]
* 8df07885 2022-02-05 misc: solve compilation warnings                   [rr- <dash@wind.garden>]
```

**Comment by Richard-L** [comment]
Okay I think I ran into something.

This is a save right before entering the room with the second Larson fight in Sanctuary:
[saveati.zip](https://github.com/rr-/Tomb1Main/files/8009164/saveati.zip)

In the NSF, load that save, wait until he begins his line "I still got a pain....", when he hasn't finished it, F5 save the game. This produces this save for me, which upon loading goes straight back to the main menu:
[save_tr1_00.zip](https://github.com/rr-/Tomb1Main/files/8009165/save_tr1_00.zip)

**Comment by rr-** [comment]
Fixed. The save should be working again without a re-save. We had an item check that the saved items should correspond to the level items, however I failed to remember that used puzzle holes can change their object numbers to puzzle done.

**Comment by rr-** [comment]
Rebased against newest master. @Richard-L here's the new version:
[Tomb1Main-2.5-35-g7286bde.zip](https://github.com/rr-/Tomb1Main/files/8011198/Tomb1Main-2.5-35-g7286bde.zip)

**Comment by walkawayy** [comment]
I'll make a checklist of what we need to test so we can keep track. Feel free to edit / add / checkmark what you tested.

- [ ]     persisting ammo, 
- [ ]     persisting medpacks
- [ ]     persisting key items
- [ ]     persisting scions
- [ ]     persisting guns
- [ ]     persisting kill count
- [ ]     persisting timer
- [ ]     persisting collected secrets
- [ ]     persisting Lara equipped guns
- [ ]     persisting Lara animations
- [ ]     persisting Lara health
- [ ]     persisting Lara air
- [ ]     not getting stationary enemies
- [ ]     dead enemies stay dead
- [ ]     water current
- [ ]     Lara falling
- [ ]     underwater
- [ ]     water surface
- [ ]     flipmaps on
- [ ]     flipmaps off
- [ ]     behavior of Bacon Lara

**Comment by Richard-L** [comment]
@rr- just like you said the previously "bugged" save now works :+1: 

Curiously after finishing the level, I started Natla's Mines with a jump over to where the push block "puzzle" is. It's a jump where you're not supposed to reach until you've changed the boat's position, but with a certain jump+shimmey manouevre you can still get over. Saving there produced this save, which upon loading prompts the main menu again:
[save_tr1_01.zip](https://github.com/rr-/Tomb1Main/files/8011302/save_tr1_01.zip)

**Comment by rr-** [comment]
I found a heavier bug with the GlobalItemReplace.

1. Save near a shotgun on slot 1
2. Pick it up
3. Save on slot 2
4. Load slot 2, Lara has shotgun, there's no shotgun on the floor
5. Load slot 1, Lara has no shotgun, but there's shotgun ammo on the floor instead of shotgun

@Richard-L could you please test if the above scenario happens on 2.5?

**Comment by Richard-L** [comment]
@rr- you mean last release v2.5 right?

As I wanted to test I jumped over to the Lost Valley shotgun, and strangely enough it shows as ammo :thinking: I hadn't noticed this at all before, but I'm also not confident I went to this spot in v2.5.
[v2.5_lost-valley_shotgun-as-ammo.zip](https://github.com/rr-/Tomb1Main/files/8011357/v2.5_lost-valley_shotgun-as-ammo.zip)

**Comment by Richard-L** [comment]
Otherwise I've tested most of the easier to test of your list @walkawayy, and I cannot find any issues at all so far.

I noticed that exploding enemies will have their explosions that hadn't finished while saving cleared upon load, but I guess this is expected behaviour.

**Comment by rr-** [comment]
Made an issue about this: #406 

Here's a new version that fixes the weapon issue:

[Tomb1Main-2.5-36-gf7c8379.zip](https://github.com/rr-/Tomb1Main/files/8011386/Tomb1Main-2.5-36-gf7c8379.zip)

Before we finalize, I would like to remove the `"flags"` key in the levels array and replace it with separate booleans for each flag. This will render the NG files from testing thus far incompatible.

**Comment by walkawayy** [comment]
> Otherwise I've tested most of the easier to test of your list @walkawayy, and I cannot find any issues at all so far.
> 
> I noticed that exploding enemies will have their explosions that hadn't finished while saving cleared upon load, but I guess this is expected behaviour.

Nice. Are you able to click the checkboxes to indicate what you tested? I think you should be able to. I can help out with some of the remaining ones later tonight.

**Comment by rr-** [comment]
It is possible that this function is limited to direct collaborators only; I invited posix to the repo as a collaborator.

**Comment by Richard-L** [comment]
Thanks @rr- . I've accepted the invitation. Can't click the checkboxes on mobile, but will try from desktop later.

**Comment by walkawayy** [comment]
I checkmarked a bunch of scenarios I tested. I tried having two different saves to be able to go back and forth forth between stuff like flip on, flip off, Bacon Lara alive/dead, etc, and I  didn't notice any issues.

**Comment by rr-** [comment]
OK, thanks. I guess I'll merge it today and then we can take care of the rest of savegame-related feature requests to test how the new sg format feels like.

**Comment by Richard-L** [comment]
Amazing guys!!

Probably a daft question, but what are "stationary enemies"?

**Comment by rr-** [comment]
Sorry I didn't explain it well; I meant enemies who'd stay inactive after loading the game, as if they never were triggered.

**Pull request merged**